File: widgets/ios/WidgetAppointment.swift
//
//  WidgetAppointment.swift
//  Schedule widget — iOS
//
//  Drop-in compatible with the shared widget JSON written by the MAUI app
//  (App Group "group.com.omnicasa.mobile", file Widgets/widget_appointments.json).
//  The CodingKeys and date format match the existing agenda widget contract.
//

import Foundation
import SwiftUI

/// One appointment as serialized by the app into the shared widget JSON.
struct WidgetAppointment: Codable, Identifiable {
    var id: Double

    let title: String?
    let subTitle: String?
    let startTime: String   // "yyyy-MM-dd HH:mm:ss"
    let endTime: String     // "yyyy-MM-dd HH:mm:ss"
    let isSuperAppointment: Bool
    let isGroupAppointment: Bool
    let isPrivate: Bool
    let isRecurrence: Bool
    let isRecurrenceException: Bool
    let backgroundColor: String?   // hex "#RRGGBB"
    let place: String?
    let imageKey: String?
    let isCommented: Bool

    enum CodingKeys: String, CodingKey {
        case id = "AppointmentId"
        case title = "Title"
        case subTitle = "SubTitle"
        case startTime = "StartTime"
        case endTime = "EndTime"
        case isSuperAppointment = "IsSuperAppointment"
        case isGroupAppointment = "IsGroupAppointment"
        case isPrivate = "IsPrivate"
        case isRecurrence = "IsRecurrence"
        case isRecurrenceException = "IsRecurrenceException"
        case imageKey = "ImageKey"
        case backgroundColor = "BackgroundColor"
        case place = "Place"
        case isCommented = "IsCommented"
    }
}

/// The shared file wrapper the app writes (`{ "Time": ..., "Appointments": [...] }`).
struct WidgetAppointmentShare: Codable {
    let time: String?
    let appointments: [WidgetAppointment]

    enum CodingKeys: String, CodingKey {
        case time = "Time"
        case appointments = "Appointments"
    }
}

extension WidgetAppointment {
    /// Parsed start, or nil if the string is malformed.
    var start: Date? { WidgetDate.parse(startTime) }

    /// Parsed end, or nil if the string is malformed.
    var end: Date? { WidgetDate.parse(endTime) }

    /// Resolved block color (hex, else a neutral accent).
    var color: Color { Color(hex: backgroundColor) ?? ScheduleWidgetTheme.accent }
}

/// Shared date parsing for the widget transport format.
enum WidgetDate {
    static let format = "yyyy-MM-dd HH:mm:ss"

    private static let formatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = format
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    static func parse(_ value: String) -> Date? { formatter.date(from: value) }
}

extension Color {
    /// Parses "#RGB", "#RRGGBB" or "#AARRGGBB"; returns nil when empty/invalid.
    init?(hex: String?) {
        guard var s = hex?.trimmingCharacters(in: .whitespaces), !s.isEmpty else { return nil }
        if s.hasPrefix("#") { s.removeFirst() }
        var value: UInt64 = 0
        guard Scanner(string: s).scanHexInt64(&value) else { return nil }

        let r, g, b, a: Double
        switch s.count {
        case 6:
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
            a = 1
        case 8:
            a = Double((value >> 24) & 0xFF) / 255
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
        case 3:
            r = Double((value >> 8) & 0xF) / 15
            g = Double((value >> 4) & 0xF) / 15
            b = Double(value & 0xF) / 15
            a = 1
        default:
            return nil
        }
        self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
    }
}

File: widgets/android/WidgetAppointment.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Omnicasa.Schedule.Widget;

/// <summary>
/// One appointment as serialized into the shared widget JSON by the app. Property names match the
/// existing agenda-widget contract (<c>Widgets/widget_appointments.json</c>) so this is drop-in.
/// </summary>
public sealed class WidgetAppointment
{
    /// <summary>Date format used in the shared JSON (<c>StartTime</c>/<c>EndTime</c>).</summary>
    public const string DateFormat = "yyyy-MM-dd HH:mm:ss";

    /// <summary>Gets or sets the stable appointment id.</summary>
    [JsonPropertyName("AppointmentId")]
    public long AppointmentId { get; set; }

    /// <summary>Gets or sets the title shown on the block.</summary>
    [JsonPropertyName("Title")]
    public string? Title { get; set; }

    /// <summary>Gets or sets the optional subtitle (e.g. place).</summary>
    [JsonPropertyName("SubTitle")]
    public string? SubTitle { get; set; }

    /// <summary>Gets or sets the start time string (<see cref="DateFormat"/>).</summary>
    [JsonPropertyName("StartTime")]
    public string? StartTime { get; set; }

    /// <summary>Gets or sets the end time string (<see cref="DateFormat"/>).</summary>
    [JsonPropertyName("EndTime")]
    public string? EndTime { get; set; }

    /// <summary>Gets or sets the block color as hex (<c>#RRGGBB</c>).</summary>
    [JsonPropertyName("BackgroundColor")]
    public string? BackgroundColor { get; set; }

    /// <summary>Gets or sets the optional place text.</summary>
    [JsonPropertyName("Place")]
    public string? Place { get; set; }

    /// <summary>Gets or sets the optional status image key.</summary>
    [JsonPropertyName("ImageKey")]
    public string? ImageKey { get; set; }

    /// <summary>Gets or sets a value indicating whether this is a super appointment.</summary>
    [JsonPropertyName("IsSuperAppointment")]
    public bool IsSuperAppointment { get; set; }

    /// <summary>Gets or sets a value indicating whether this is a group appointment.</summary>
    [JsonPropertyName("IsGroupAppointment")]
    public bool IsGroupAppointment { get; set; }

    /// <summary>Gets or sets a value indicating whether this is private.</summary>
    [JsonPropertyName("IsPrivate")]
    public bool IsPrivate { get; set; }

    /// <summary>Gets or sets a value indicating whether this is a recurrence.</summary>
    [JsonPropertyName("IsRecurrence")]
    public bool IsRecurrence { get; set; }

    /// <summary>Gets or sets a value indicating whether this is a recurrence exception.</summary>
    [JsonPropertyName("IsRecurrenceException")]
    public bool IsRecurrenceException { get; set; }

    /// <summary>Gets or sets a value indicating whether this has a comment.</summary>
    [JsonPropertyName("IsCommented")]
    public bool IsCommented { get; set; }

    /// <summary>Gets the parsed start, or null when the string is missing/malformed.</summary>
    public DateTime? Start => Parse(StartTime);

    /// <summary>Gets the parsed end, or null when the string is missing/malformed.</summary>
    public DateTime? End => Parse(EndTime);

    private static DateTime? Parse(string? value)
        => DateTime.TryParseExact(value, DateFormat, CultureInfo.InvariantCulture, DateTimeStyles.None, out var d)
            ? d
            : null;
}

/// <summary>The shared-file wrapper the app writes: <c>{ "Time": ..., "Appointments": [...] }</c>.</summary>
public sealed class WidgetAppointmentShare
{
    /// <summary>Gets or sets the snapshot time string.</summary>
    [JsonPropertyName("Time")]
    public string? Time { get; set; }

    /// <summary>Gets or sets the appointments in the snapshot.</summary>
    [JsonPropertyName("Appointments")]
    public List<WidgetAppointment> Appointments { get; set; } = new List<WidgetAppointment>();

    /// <summary>Deserializes a share payload; returns an empty list on any failure.</summary>
    /// <param name="json">The shared JSON.</param>
    /// <returns>The parsed appointments, or an empty list.</returns>
    public static IReadOnlyList<WidgetAppointment> Read(string json)
    {
        try
        {
            return JsonSerializer.Deserialize<WidgetAppointmentShare>(json)?.Appointments
                   ?? new List<WidgetAppointment>();
        }
        catch (JsonException)
        {
            return new List<WidgetAppointment>();
        }
    }
}
